glean-core/src/upload/mod.rs

```diff
@@ -0,0 +1,357 @@
+}
+
+/// Attempt to delete a file.
+///
+/// This won't panic if not able to delete, just log.
+fn delete_file(path: &Path) {
+    match remove_file(path) {
+        Err(e) => log::error!("Error deleting file {}. {}", path.display(), e),
+        _ => log::info!("Deleted file {}", path.display()),
+    };
+}
+
+/// Get the file name from a path as a &str.
+///
+/// This won't panic if not able to get file name, just log.
+fn get_file_name_as_str(path: &Path) -> Option<&str> {
+    match path.file_name() {
+        None => {
+            log::warn!("Error getting file name from path: {}", path.display());
+            None
+        }
+        Some(file_name) => {
+            let file_name = file_name.to_str();
+            if file_name.is_none() {
+                log::warn!("File name is not valid unicode: {}", path.display());
+            }
+            file_name
+        }
+    }
+}
+
+/// Creates requests for ping upload and process responses from them.
+pub struct PingUploadHelper {
+    /// A queue of ping UUIDs to upload
+    queue: PingsQueue,
+    /// The path to the pending_pings directory
+    pending_pings_dir: PathBuf,
+}
+
+impl PingUploadHelper {
+    /// Create a new PingUploadHelper.
+    ///
+    /// Starts an empty queue to be filled with the pending pings
+    /// file paths that should be processed and uploaded next.
+    pub fn new(data_path: &Path) -> Self {
+        Self {
+            queue: PingsQueue::new(),
+            pending_pings_dir: data_path.join("pending_pings"),
+        }
+    }
+
+    /// Creates a `PingRequest` for the first possible ping in the queue.
+    ///
+    /// If one of the files is malformed or generates errors, it will
+    /// keep trying while there are still pending pings.
+    pub fn create_next_ping_request(&self) -> Option<PingRequest> {
+        let mut ping = self.get_next_ping();
+        // As long as there are pings in the queue this will
+        // keep trying to create a `PingRequest` from it.
+        while ping.is_some() {
+            // Safe unwrap, we state that ping "is_some" before entering the loop
+            let (uuid, path) = ping.unwrap();
+            let request = self.process_ping_file(uuid, path);
+            if request.is_some() {
+                return request;
+            }
+            ping = self.get_next_ping();
+        }
+        None
+    }
+
+    /// Process the next ping, if there are any in the queue.
+    ///
+    /// Will try **once** to re-process the pending_pings folder
+    /// in case the queue is empty.
+    fn get_next_ping(&self) -> Option<(Uuid, PathBuf)> {
+        let mut attempts = 0;
+        loop {
+            if let Some(uuid) = self.queue.process() {
+                // Create the ping file path from the UUID
+                let path = self.pending_pings_dir.join(uuid.to_string());
+                return Some((uuid, path));
+            }
+
+            if let Err(e) = self.process_pings_dir() {
```

Comment:
What about this:

- only to the `process_pings_dir` when this module is initialized/built
- do that off the main thread
- every time a new ping is created (outside of this module), in addition to storing it to disk, also add it to the queue.

This way we would not require to process the pings dir when the queue is empty and we'd guarantee that the full processing only happens once, at startup, off the main thread.

Reply:
Yes, I agree that would be ideal. I thought of something similar, but didn't want to digress too much from the current way things work for this PR. I tried to "copy" as much as I could from the Android implementation. 

Would you agree to make this change in a future PR, specifically the one where I add interface for the upload module to the Glean object? I think that would make sense, since this change would require that the upload helper is initialized upon initialization of the Glean object.

Reply:
Note that the current Android implementation already runs off the main thread, since it runs on a thread owned by the WorkManager. If we were to enable this in the current form, we would see a performance regression due to disk I/O.

Reply:
Talked to @badboy about this and we think the best approach would be to do what you said and spawn a new thread on initialization to process de directory, but also also process the files and save Request objects to the queue so that we only do disk I/O on this step.

Reply:
I did this, it works but it resulted in a lot of changes in the structure of the code.